File: solutions/problem_1504_count_submatrices_with_all_ones/solution.py
"""
Problem 1504: Count Submatrices With All Ones
===========================================
Difficulty: Medium
Tags: Array, Dynamic Programming, Stack, Matrix, Monotonic Stack
Pattern: Monotonic Stack

Time Complexity:  O(n)
Space Complexity: O(n)
"""

class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        # Monotonic stack - O(n) time, O(n) space
        n = len(mat)
        result = [0] * n
        stack = []  # indices
        for i in range(n):
            while stack and mat[i] > mat[stack[-1]]:
                idx = stack.pop()
                result[idx] = i - idx
            stack.append(i)
        return result

File: solutions/problem_1504_count_submatrices_with_all_ones/solution.cpp
/*
 * Problem 1504: Count Submatrices With All Ones
 * ===========================================
 * Difficulty: Medium
 * Tags: Array, Dynamic Programming, Stack, Matrix, Monotonic Stack
 * Pattern: Monotonic Stack
 *
 * Time Complexity:  O(n)
 * Space Complexity: O(n)
 */

#include <stack>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        // Monotonic stack - O(n) time, O(n) space
        int n = mat.size();
        vector<int> result(n, 0);
        stack<int> st;
        for (int i = 0; i < n; i++) {
            while (!st.empty() && mat[i] > mat[st.top()]) {
                int idx = st.top(); st.pop();
                result[idx] = i - idx;
            }
            st.push(i);
        }
        return result;
    }
};
